File: GW3/LogFileHandling/Query/Tokens/TokenWhere.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenWhere : Token
    {
        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();

            return parser.PeekString().ToLower() == "where";
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            parser.NextString();
            return new TokenWhere();
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenWhere.ts
class TokenWhere extends Token {

    public tokenType: TokenType = TokenType.TokenWhere;

    public proposals: SuggestionInterface[] = <SuggestionInterface[]>[{ suggestion: "where" }];

    public canBeUsed(parser: QueryParser): boolean {
        parser.skipSpaces();
        return parser.peekString().toLowerCase() == "where";
    }
    public extract(parser: QueryParser): Token {
        parser.skipSpaces();
        parser.nextString();
        let token = new TokenWhere();
        token.value = "where";
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[] {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined') {
            switch (nextToken.tokenType) {
                case TokenType.TokenName:
                    return [new TokenCompare()];
                case TokenType.TokenOpenParenthesis:
                    return [new TokenName(), new TokenOpenParenthesis()];
            }
        }
        return [];
    }
}
